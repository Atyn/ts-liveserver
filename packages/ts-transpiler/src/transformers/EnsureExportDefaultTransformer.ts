import TypeScript from 'typescript'

/*
Replaces process.env.NODE_ENV in code
*/
export default class EnsureExportDefaultTransformer
	implements TypeScript.CustomTransformer
{
	private context: TypeScript.TransformationContext
	constructor(context: TypeScript.TransformationContext) {
		this.context = context
	}
	transformSourceFile(
		sourceFile: TypeScript.SourceFile,
	): TypeScript.SourceFile {
		if (this.hasDefaultExport(sourceFile)) {
			return sourceFile
		}
		return TypeScript.factory.updateSourceFile(sourceFile, [
			...sourceFile.statements,
			TypeScript.factory.createExportAssignment(
				undefined,
				undefined,
				undefined,
				TypeScript.factory.createNull(),
			),
		])
	}
	transformBundle(node: TypeScript.Bundle): TypeScript.Bundle {
		return node
	}
	private hasDefaultExport(sourceFile: TypeScript.SourceFile): boolean {
		for (const statement of sourceFile.statements) {
			if (TypeScript.isExportAssignment(statement)) {
				return true
			} else if (
				TypeScript.isExportDeclaration(statement) &&
				statement.exportClause &&
				TypeScript.isNamedExports(statement.exportClause) &&
				statement.exportClause.elements &&
				statement.exportClause.elements.length &&
				statement.exportClause.elements.some(
					(element) => element.name.text === 'default',
				)
			) {
				return true
			} else if (
				statement.modifiers &&
				statement.modifiers.some(
					(node) => node.kind === TypeScript.SyntaxKind.ExportKeyword,
				) &&
				statement.modifiers.some(
					(node) => node.kind === TypeScript.SyntaxKind.DefaultKeyword,
				)
			) {
				return true
			}
		}
		return false
	}
}
