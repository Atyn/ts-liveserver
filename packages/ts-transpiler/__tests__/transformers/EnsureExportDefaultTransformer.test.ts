import EnsureExportDefaultTransformer from '../../src/transformers/EnsureExportDefaultTransformer'
import TypeScript from 'typescript'

describe('EnsureExportDefaultTransformer', () => {
	it('Should have default export redirects', async () => {
		const input = 'export default interface Hello { yo: string }'
		expect(await transformWithPlugin(input)).toMatchSnapshot()
	})
	it('Should not add default export', async () => {
		const input = 'export default class Hello { yo: string }'
		expect(await transformWithPlugin(input)).toMatchSnapshot()
	})
	it('Should not add default export', async () => {
		const input = 'export { a as default }'
		expect(await transformWithPlugin(input)).toMatchSnapshot()
	})
	it('Should not add default export', async () => {
		const input = 'export default hello'
		expect(await transformWithPlugin(input)).toMatchSnapshot()
	})
})

const compilerOptions: TypeScript.CompilerOptions = {
	allowJs: true,
	jsx: TypeScript.JsxEmit.React,
	checkJs: false,
	noResolve: true,
	esModuleInterop: true,
	skipLibCheck: true,
	pretty: true,
	allowUnreachableCode: true,
	target: TypeScript.ScriptTarget.ES2020,
	declaration: false,
	module: TypeScript.ModuleKind.ES2020,
	moduleResolution: TypeScript.ModuleResolutionKind.NodeJs,
	sourceMap: false,
}

const transformers: TypeScript.CustomTransformers = {
	after: [(context) => new EnsureExportDefaultTransformer(context)],
}

async function transformWithPlugin(code: string): Promise<string> {
	const results = await TypeScript.transpileModule(code, {
		compilerOptions: compilerOptions,
		fileName: 'hello.ts',
		transformers: transformers,
	})
	return results.outputText.trim()
}

async function transformWithoutPlugin(code: string): Promise<string> {
	const results = await TypeScript.transpileModule(code, {
		compilerOptions: compilerOptions,
		fileName: 'hello.ts',
	})
	return results.outputText.trim()
}
