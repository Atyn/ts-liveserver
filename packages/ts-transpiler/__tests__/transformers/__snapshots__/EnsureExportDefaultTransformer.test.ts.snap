// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnsureExportDefaultTransformer Should have default export redirects 1`] = `
"export {};
export default null;"
`;

exports[`EnsureExportDefaultTransformer Should not add default export 1`] = `
"export default class Hello {
}"
`;

exports[`EnsureExportDefaultTransformer Should not add default export 2`] = `"export { a as default };"`;

exports[`EnsureExportDefaultTransformer Should not add default export 3`] = `"export default hello;"`;
